golem_cli/src/service.rs

```diff
@@ -0,0 +1,132 @@
+use anyhow::{bail, Context, Result};
+use futures::{future::FutureExt, select};
+use std::{io, process::Stdio};
+use tokio::{io::AsyncBufReadExt, net::TcpStream, process::Command, time};
+use url::Url;
+use ya_service_api_web::{DEFAULT_YAGNA_API_URL, YAGNA_API_URL_ENV_VAR};
+
+async fn wait_for_socket(addr: std::net::SocketAddr) -> Result<()> {
+    let mut timeout_remaining = 3;
+
+    loop {
+        match TcpStream::connect(addr).await {
+            Ok(_) => {
+                log::info!("socket found, addr: {}", addr);
+                return Ok(());
+            }
+            Err(err) => match err.kind() {
+                io::ErrorKind::ConnectionRefused => {
+                    log::info!("Waiting for socket ...");
+                    if timeout_remaining > 0 {
+                        time::delay_for(time::Duration::from_secs(1)).await;
+                        timeout_remaining -= 1;
+                    } else {
+                        bail!("Could not connect to the bus socket");
+                    }
+                }
+                _ => break Err(err.into()),
+            },
+        }
+    }
+}
+
+async fn reader_to_log<T: tokio::io::AsyncRead + Unpin>(name: String, reader: T) {
+    let mut reader = tokio::io::BufReader::new(reader);
+    let mut buf = Vec::new();
+    loop {
+        match reader.read_until(b'\n', &mut buf).await {
+            Ok(len) => {
+                if len > 0 {
+                    log::debug!(
```

Comment:
I'd suggest redirecting daemon logs to a file and displaying provider logs with their respective log level (preserving colored outputs)

Reply:
There's a `TODO` for this in line 67

Reply:
Yes, and it's rather vague :slightly_smiling_face: